### PR TITLE
Feature/12 dogu is starting page

### LINKED ISSUE #12: Show the dogu is starting page when a dogu is starting.

### PR DESCRIPTION
**Pull request by Subaro**
Resolves #12

### CODE REVIEW

File: controllers/ingressUpdater.go [hunk 1]
```diff
@@ -80,10 +112,47 @@ func (i *ingressUpdater) UpdateIngressOfService(ctx context.Context, service *co
...
+			if err != nil {
+				log.FromContext(ctx).Error(fmt.Errorf("failed to wait for the readiness of the [%s] deployment: %w", service.GetName(), err), "failed to update ingress object")
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (i *ingressUpdater) updateIngressObject(ctx context.Context, cesService CesService, service *corev1.Service, isMaintenanceMode bool) error {
```
Comment (cesmarvin) [inline controllers/ingressUpdater.go:149]: 🚫 **[checkstyle]** <funlen> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Function 'updateIngressObject' is too long (66 > 60)

File: controllers/serviceController_int_test.go [hunk 2]
```diff
@@ -7,6 +7,11 @@ import (
 	"context"
 	_ "embed"
 	"fmt"
+	"github.com/cloudogu/k8s-service-discovery/controllers/dogustart"
+	v12 "k8s.io/api/apps/v1"
```
Comment (nhinze23) [inline controllers/serviceController_int_test.go:11]: delete naming

File: controllers/ingressUpdater.go [hunk 3]
```diff
@@ -4,21 +4,29 @@ import (
...
-	staticContentBackendRewrite    = "/errors/503.html"
-	ingressRewriteTargetAnnotation = "nginx.ingress.kubernetes.io/rewrite-target"
+	staticContentBackendName           = "nginx-static"
+	staticContentBackendPort           = 80
+	staticContentBackendRewrite        = "/errors/503.html"
+	staticContentDoguIsStartingRewrite = "/errors/starting.html"
+	ingressRewriteTargetAnnotation     = "nginx.ingress.kubernetes.io/rewrite-target"
 )
 
+const waitForDeploymentTimeout = time.Minute * 30
```
Comment (cesmarvin) [inline controllers/ingressUpdater.go:27]: 🚫 **[checkstyle]** <deadcode> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>`waitForDeploymentTimeout` is unused

File: controllers/ingressUpdater.go [hunk 4]
```diff
@@ -4,21 +4,29 @@ import (
...
+const waitForDeploymentTickRate = time.Second
```
Comment (cesmarvin) [inline controllers/ingressUpdater.go:28]: 🚫 **[checkstyle]** <deadcode> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>`waitForDeploymentTickRate` is unused

**Comment by nhinze23** [comment]
It would be nice if you additionally check the dogu resource on creation and set the starting page at this moment as well.

File: controllers/serviceController_int_test.go [hunk 5]
```diff
@@ -35,7 +40,9 @@ var serviceWebApp = &corev1.Service{}
 var serviceAdditionalBytes []byte
 var serviceAdditional = &corev1.Service{}
 
-func init() {
+var serviceDeployment = &v1.Deployment{}
+
+func resetData() {
```
Comment (ppxl) [inline controllers/serviceController_int_test.go:45]: The function name and the belonging package variable are cues that tests _may be_ interdependent on shared data. A better solution would include data and parse them per tests. This way no data are shared and no dependency is built.

File: controllers/serviceController_int_test.go [hunk 6]
```diff
@@ -50,34 +57,65 @@ func init() {
...
-			expectedIngress := &networking.IngressList{}
-
-			Eventually(func() bool {
-				err := k8sClient.List(ctx, expectedIngress)
-				return err == nil
-			}, timeoutInterval, pollingInterval).Should(BeTrue())
-
-			Expect(len(expectedIngress.Items)).Should(Equal(0))
-		})
+		//
```
Comment (ppxl) [inline controllers/serviceController_int_test.go:95]: What happens with this dead code?

File: controllers/serviceController_int_test.go [hunk 7]
```diff
@@ -127,18 +176,63 @@ var _ = Describe("Creating ingress objects with the ingress generator", func() {
...
-			Expect(expectedIngress.Items[1].Name).Should(Equal("nexus-docker-registry"))
-			Expect(*expectedIngress.Items[1].Spec.IngressClassName).Should(Equal(myIngressClassName))
-			Expect(expectedIngress.Items[1].Spec.Rules[0].HTTP.Paths[0].Path).Should(Equal("/v2"))
-			Expect(*expectedIngress.Items[1].Spec.Rules[0].HTTP.Paths[0].PathType).Should(Equal(networking.PathTypePrefix))
-			Expect(expectedIngress.Items[1].Spec.Rules[0].HTTP.Paths[0].Backend.Service.Name).Should(Equal("nexus"))
-			Expect(expectedIngress.Items[1].Spec.Rules[0].HTTP.Paths[0].Backend.Service.Port.Number).Should(Equal(int32(8082)))
-			Expect(expectedIngress.Items[1].Annotations[ingressRewriteTargetAnnotation]).Should(Equal("/nexus/repository/docker-registry/v2"))
 		})
+
+		//It("Should create ingress object for multiple ces services", func() {
```
Comment (ppxl) [inline controllers/serviceController_int_test.go:183]: likewise: deadcode

**Review by ppxl — DISMISSED** [review]
Looks good most of the time. Good work! I found some minor things. The largest question marks lingers around the point of possibly unnecessary K8s permissions.

File: controllers/dogustart/doguHealthChecker.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	"k8s.io/client-go/kubernetes"
+	"time"
+)
+
+type deploymentReadyChecker struct {
+	namespace string
+	client    kubernetes.Interface
+}
+
+// WaitOptions defines the timeout and tickrate for waiting functions.
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/dogustart/doguHealthChecker.go:17]: I expanded the comment.

File: controllers/dogustart/doguHealthChecker.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+// WaitForReady allows the execution of code when the deployment switches from the not ready state into the ready state.
+func (d *deploymentReadyChecker) WaitForReady(ctx context.Context, deploymentName string, waitOptions WaitOptions, onReady func(ctx context.Context)) error {
+	ok, err := d.IsReady(ctx, deploymentName)
+	if err != nil {
+		return err
+	}
+
+	if ok {
+		// pod is already ready
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/dogustart/doguHealthChecker.go:51]: I removed it.

File: controllers/dogustart/doguHealthChecker.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		return err
+	}
+
+	closeTimeout := time.After(waitOptions.Timeout)
+	for {
+		select {
+		case <-closeTimeout:
+			watch.Stop()
+			return fmt.Errorf("failed to wait for deployment readiness: timeout after [%s] while waiting of pod being ready", waitOptions.Timeout.String())
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/dogustart/doguHealthChecker.go:68]: The `String()` method always delivers a unit, e.g., 20s or 4m

File: controllers/serviceController_int_test.go (hunk 5 above)
Comment (Subaro) [reply controllers/serviceController_int_test.go:45]: The resetData should be called for every test.

File: controllers/serviceController_int_test.go (hunk 6 above)
Comment (Subaro) [reply controllers/serviceController_int_test.go:95]: I commented them out for faster feedback. I commented them back in.

File: controllers/serviceController_int_test.go (hunk 7 above)
Comment (Subaro) [reply controllers/serviceController_int_test.go:183]: I commented them out for faster feedback. I commented them back in.

File: controllers/serviceController_int_test.go (hunk 6 above)
Comment (Subaro) [reply controllers/serviceController_int_test.go:95]: They also work :)

File: config/rbac/role.yaml [hunk 11]
```diff
@@ -10,6 +10,7 @@ rules:
       - ingresses
       - services
       - secrets
+      - deployments
```
[the remark this answers is not in the document]
Comment (Subaro) [reply config/rbac/role.yaml:13]: I am not sure. I think we need to perform a small audit to fully understand which permissions are required for the service discovery.

[the remark this answers is not in the document]
Comment (Subaro) [reply config/rbac/role.yaml]: We reworked the the permissions for the service discovery.

**Review by jelemux — APPROVED** [review]
LGTM!

**Pull request merged**